src/OAuthClient.php

```diff
@@ -2,403 +2,483 @@
 
-        if (isset($data['code'])) {
-            return $data['code'];
+        return $url;
+    }
+
+    /**
+     * Get the oAuth2 state variable for security checks
+     * @return string
+     */
+    public function getState(): string
+    {
+        return $this->state;
+    }
+
+    /**
+     * Parse Grant Token from URL
+     *
+     * @param UriInterface $uri
+     * @return string|null
+     */
+    public static function parseGrantTokenFromUrl(UriInterface $uri): ?string
+    {
+        parse_str($uri->getQuery(), $output);
+        if (isset($output['code'])) {
+            return $output['code'];
         }
 
         return null;
     }
 
-    public function getRegion(): Region
+    /**
+     * Get the Zoho Region
+     *
+     * @return string
+     */
+    public function getRegion(): string
     {
         return $this->region;
     }
 
-    public function setRegion(Region $region): self
+    /**
+     * Set the Zoho Region
+     *
+     * @param string $region
+     * @return self
+     */
+    public function setRegion(string $region): self
     {
         $this->region = $region;
+        $this->createProvider();
 
         return $this;
     }
 
+    /**
+     * Set the Grant Code
+     *
+     * @param string $grantCode
+     * @return self
+     */
     public function setGrantCode(string $grantCode): self
     {
         $this->grantCode = $grantCode;
 
         return $this;
     }
 
-    public function setState(string $state): self
-    {
-        $this->state = $state;
-
-        return $this;
-    }
-
+    /**
+     * Set the Scopes
+     *
+     * @param array $scopes
+     * @return self
+     */
     public function setScopes(array $scopes): self
     {
         $this->scopes = $scopes;
 
         return $this;
     }
 
+    /**
+     * Set Offline Mode
+     *
+     * @return self
+     */
     public function onlineMode(): self
     {
         $this->offlineMode = false;
 
         return $this;
     }
 
+    /**
+     * Check if is offline
+     *
+     * @return bool
+     */
     public function isOffline(): bool
     {
         return $this->offlineMode;
     }
 
+    /**
+     * Check is is online
+     *
+     * @return bool
+     */
     public function isOnline(): bool
     {
         return ! $this->offlineMode;
     }
 
+    /**
+     * Set the redirect uri
+     *
+     * @param string $redirectUri
+     * @return self
+     */
     public function setRedirectUri(string $redirectUri): self
     {
         $this->redirectUri = $redirectUri;
+        $this->createProvider();
 
         return $this;
     }
 
-    public function useCache(Cache\CacheItemPoolInterface $cacheItemPool): self
+    /**
+     * Set the cache pool to use
+     *
+     * @param Cache\CacheItemPoolInterface $cacheItemPool
+     * @param string $cachePrefix
+     * @return self
+     */
+    public function useCache(Cache\CacheItemPoolInterface $cacheItemPool, string $cachePrefix = 'zoho_crm_'): self
```

Comment:
There's a parameter for cache prefix here but it's never used.

Reply:
Good catch, fixed